### ISSUE TITLE
On the new terminology and making testable claims

### ISSUE THREAD

**Issue by AngusG** [anishathalye/obfuscated-gradients#5]
Thank you for this useful work that will help the community be more rigorous when claiming robustness to adversarial examples. 

Some of the new terminology (obfuscated gradients) could interfere with the purpose of the paper, which (my wording) is to "demystify" common reasons that defense researchers frequently falsely conclude robustness. I find _gradient masking_ to be a sufficiently general term to cover all scenarios in the paper, and the use of new terminology is likely to confuse researchers into thinking there is some practical difference between them, which I don't believe there is. 

Disclosure: After further experiments, an element of gradient masking also affected my stochastic binarized neural networks in my [ICLR submission](https://openreview.net/forum?id=HkTEFfZRb) which led to especially good performance against the Carlini & Wagner L2 attack. That said, low-precision does seem to provide a unique regularisation effect which I am actively working to better characterise, and it is both interesting and encouraging that adversarial training works equally well with BNNs given their significantly reduced capacity (using the term loosely), which Madry et al. argue we ought to have _more_ of.

I would also like to point out that "Backward Pass Differentiable Approximation" is the well known [Straight Through Estimator ](https://arxiv.org/abs/1308.3432) for back-propagating through non-differentiable nodes which wasn't cited. 

Lastly, what are your thoughts on adding a fourth item, along with distance metric, perturbation bound, robustness %, for verifying claims: **clean test accuracy**.

When you tested Stochastic Activation Pruning in Section 5.3.1, you mention:

> We follow the authors advice and choose r so the test accuracy drops by only 5%.

What if they told you to drop accuracy by 10 or 20%? Where do you draw the line? As far as I can tell the community has never attempted to define what a reasonable drop in clean test accuracy is when reporting robustness to adversarial examples. Typically, people start with something state-of-the-art, and then claim some small drop in accuracy such that they see a boost against adversarial examples. Clearly this approach of attempting to defend so called state-of-the-art models isn't working. Heavily regularising a vanilla CNN (e.g with weight decay, inspired by [Tanay & Griffin](https://arxiv.org/abs/1608.07690)) is competitive with data augmentation / adversarial training techniques, but you see significant degradation of clean test accuracy as well. Unlike with adversarial training however, there is reason to believe that a model regularised in this way will be robust to a greater _variety_ of attacks, and the fall-off less severe immediately beyond the perturbation magnitude used in training, as shown in Figure 6 in [Madry et al](https://arxiv.org/pdf/1706.06083.pdf).

The relationship between test accuracy and adversarial accuracy was shown more formally on a synthetic data set in [Adversarial Spheres](https://openreview.net/forum?id=SyUkxxZ0b), where driving clean test error down to zero causes the distance to nearest error to become small with O(1/sqrt(d)).

My view is that the adversarial phenomenon has shown that "state-of-the-art" scores are inflated, not that underfitting models are weak or lacking expressiveness. Why should we place more emphasis on the test set (**one** subset of the population), than the infinite number of other equally valid subsets of adversarial examples. 

I believe that one way forward, without relying on heavy data augmentation, is to start with small models that learn _something_ useful, test that whatever little accuracy they obtain does not degrade with adversarial and non-examples, then progressively add more capacity in an iterative loop until satisfactory performance is reached. 

Hope to discuss more.

**Comment by carlini** [comment]
> I find gradient masking to be a sufficiently general term

See #2 for a long discussion of this.

> I would also like to point out that "Backward Pass Differentiable Approximation" is the well known Straight Through Estimator for back-propagating through non-differentiable nodes which wasn't cited.

Thanks for pointing us at this. I was unaware of this paper -- it does indeed look like we've re-invented this concept (or something very close to it) and then given it a different name.

> As far as I can tell the community has never attempted to define what a reasonable drop in clean test accuracy is when reporting robustness to adversarial examples. 

I definitely agree. This is a general problem in any context where there's an accuracy-vs-robustness tradeoff, and I don't know if there is any right answer. I personally believe (with no justification at all -- I'd be happy to be wrong here!) that we will need to accept at least 5-10% on some datasets if we want robustness. Even on MNIST, there are plenty of images that have dubious labels where classifying them "correctly" almost implies adversarial examples exist.

We should also probably revise that sentence because the authors don't explicitly say "choose so that it is 5% worse" but instead the model that performs 5% worse on clean data is the one they find is most accurate on adversarial data. (For our attacks in particular, the exact number we pick isn't all that important. That is, if we increase the drop probability so model accuracy is 20% worse, our attack results are still nearly identical)

> Why should we place more emphasis on the test set (one subset of the population), than the infinite number of other equally valid subsets of adversarial examples.

This is definitely a good point. The only reason we (and probably others) use the test set is for convenience. It's the same with why we use l_infinnity or l_2 distance to measure image similarity. There probably are better things to do, but given that models are vulnerable even when only considering the test data, and even only under l_2 or l_infinity distances, it's a reasonable approximation for the time being. When adversarial examples become solved on the test set under these simple distance metrics, we will have to start constructing more complicated analyses -- and indeed as you observe others have started that already, which is great.

**Comment by AngusG** [comment]
I agree that we will have to experience a drop in clean test accuracy to achieve robustness.  Also to clarify my comment about "Why should we place emphasis on the test set" -- I want to be clear that I'm talking about the clean, aka non-perturbed test set vs evaluating performance on adversarial examples as **standard**. For example, I don't view infinity norm epsilon=8/255 bounded perturbations for CIFAR-10 to be "worst case" examples, I view these as equally valid examples. When using the "random-start" Madry et al. approach, these examples are much more representative of a true held out "test set" since they technically change during every evaluation, therefore it's harder to overfit them. It's unlikely that you are going to test so many times that you memorize the entire epsilon ball, whereas its very likely that you can memorize enough statistics to "generalize" on the single clean test set.

Also, here's some [evidence](https://arxiv.org/pdf/1802.04457.pdf) for the argument I made earlier about small, well regularized CNNs. As far as I know, but i'm happy to be presented with evidence otherwise, this is the first compelling defense against the [Nguyen et al. (2015)](https://arxiv.org/abs/1412.1897) _fooling images_ or ``rubbish examples'' from Goodfellow et al. (2015), other than RBF networks which we know don't generalize/scale well.

**Comment by anishathalye** [comment]
Nicholas talked to Ian and others about this, and I think the terminology issue has been resolved.
